### PR TITLE
[https://nvbugs/6284101][fix] Fix guided decoding for Harmony parser

### PR DESCRIPTION
**Pull request by dongfengy**
## Dev Engineer Review

- Adds model-aware Harmony guided-decoding adaptation for raw Python `LLM.generate()`.
- Applies JSON, regex, and grammar constraints after the Harmony final-channel marker.
- Shares guided-decoding adaptation with the serving path.
- Adds parser-aware final-content extraction and JSON-mode scoring.
- Preserves behavior for normal reasoning parsers, llguidance, plain models, and unsupported guides.
- Adds parsing fallbacks for failures and empty parser output.
- Adds CPU test-list coverage and removes the GPT-OSS/Eagle3 waiver.
- No configuration typos or unintended scope changes found.

## QA Engineer Review

- Adds generation-content tests for raw text, Qwen3, non-Harmony parsers, token IDs, and explicit Harmony parsing.
- Adds sampling-parameter tests for Harmony adaptation, JSON schemas, parser resolution, unsupported guides, plain models, structural tags, and reasoning-parser sequences.
- Adds `unittest/evaluate/test_generation_content.py` to `tests/integration/test_lists/l0_cpu.yml`.
- Removes `TestGPTOSS::test_eagle3_guided_decoding_4gpus[one_model]` from `tests/integration/test_lists/waives.txt`.
- Test-list entries have valid paths and no reported duplicates.
- Verdict: sufficient.

## Description

Fix guided decoding for the raw Python `LLM.generate()` path when GPT-OSS uses the Harmony protocol with xgrammar. Caller-provided JSON, regex, and grammar constraints are enriched with model-aware Harmony framing so they apply only after the final-channel marker instead of constraining the reasoning transcript.

The serving path keeps its existing behavior while sharing the same reasoning-parser adapter. JSON-mode evaluation now scores parser-selected final content, covering plain output, explicit Harmony output, and explicit normal reasoning-parser output without guessing a parser for unrelated models.

Raw guided-decoding behavior remains unchanged for normal reasoning parsers and llguidance. This PR does not change the public API.

## Test Coverage

- Post-rebase pre-commit on all changed files: passed on exact head `059d1c9a76`.
- Incremental CUDA 13.2/TensorRT 10.16 SM90+SM100 rebuild on exact head `059d1c9a76`: passed.
- Focused guide-adaptation/final-content units: 46 passed; strict tool-guided construction: 8 passed.
- Broad affected unit/API/bindings suite: 185 passed, with the same three unrelated upstream Qwen3 auto-detection failures from missing logger methods.
- Qwen3 raw sync+async: passed with xgrammar and llguidance; Qwen3 100-sample JSON evaluation: 77/100.
- Llama-3.1-8B guided decoding: xgrammar and llguidance both passed (2 passed, 74/100 each).
- Exact four-B200 GPT-OSS/Eagle3 regression on `059d1c9a76`: passed with JSON accuracy 78/100 and grammar accuracy 96/100 (threshold 50.369).

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/evaluate/interface.py [hunk 1]
```diff
@@ -66,6 +68,47 @@ def get_model_context(llm: Any) -> tuple[str, str]:
...
+            return text
+        return text
+
+    try:
+        # Normal reasoning formats expose final content through their parser.
+        parser = ReasoningParserFactory.create_reasoning_parser(
+            reasoning_parser)
+        return parser.parse(text).content or text
+    except Exception:
+        return text
```
[the remark this answers is not in the document]
Comment (dongfengy) [reply tensorrt_llm/evaluate/interface.py:110]: Addressed in 059d1c9a76: narrowed the Harmony and normal-parser fallback catches to the expected import, type, runtime, and parsing failures.

File: tensorrt_llm/evaluate/json_mode_eval.py [hunk 2]
```diff
@@ -24,9 +24,28 @@
...
+        return json.loads(text)
+    except json.JSONDecodeError as original_error:
+        # Reasoning models may return a raw transcript such as Harmony channels
+        # or `<think>...</think>{...}`. Score the final answer content when it
+        # can be extracted; otherwise preserve the original JSON failure.
+        final_content = extract_final_content_from_generation(
+            output, reasoning_parser=reasoning_parser)
+        if final_content == text:
+            raise original_error
+        return json.loads(final_content)
```
[the remark this answers is not in the document]
Comment (dongfengy) [reply tensorrt_llm/evaluate/json_mode_eval.py]: Addressed in 059d1c9a76: added a local JsonValue alias and explicit return annotation for _load_json_from_generation.

File: tests/unittest/evaluate/test_generation_content.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (dongfengy) [reply tests/unittest/evaluate/test_generation_content.py:2]: Addressed in 059d1c9a76: restored the repository-standard NVIDIA Apache-2.0 license header on the new test file.

File: tests/unittest/evaluate/test_generation_content.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    harmony_module = types.ModuleType("tensorrt_llm.serve.harmony_adapter")
+
+    def _unexpected_adapter_lookup():
+        raise AssertionError("Harmony must be selected from model context")
+
+    harmony_module.get_harmony_adapter = _unexpected_adapter_lookup
+    monkeypatch.setitem(sys.modules, "tensorrt_llm.serve.harmony_adapter", harmony_module)
+    output = _request_output("not json", token_ids=[1, 2, 3])
+
+    assert extract_final_content_from_generation(output) == "not json"
```
[the remark this answers is not in the document]
Comment (dongfengy) [reply tests/unittest/evaluate/test_generation_content.py:70]: Addressed in 059d1c9a76: added explicit qwen3 plus token_ids coverage, with Harmony adapter lookup patched to fail if that non-Harmony path touches it.

File: tests/unittest/evaluate/test_generation_content.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import sys
+import types
+from types import SimpleNamespace
+
+from tensorrt_llm.evaluate.interface import extract_final_content_from_generation
+from tensorrt_llm.llmapi.reasoning_parser import HARMONY_REASONING_PARSER
```
Comment (brnguyen2) [inline tests/unittest/evaluate/test_generation_content.py:20]: This is the first file in a new `tests/unittest/evaluate/` directory, and pre-merge CI only runs unit tests listed in `tests/integration/test_lists/test-db/l0_*.yml` — no list references this directory, so these tests will never execute. The llmapi additions are covered only because `unittest/llmapi/test_sampling_params.py` is already in `l0_cpu.yml:59`. Please add `unittest/evaluate/test_generation_content.py` to `l0_cpu.yml` (no GPU needed) alongside it.

File: tensorrt_llm/llmapi/llm.py [hunk 6]
```diff
@@ -1390,6 +1393,24 @@ def _prepare_sampling_params(
...
+                        getattr(self._hf_model_config, "model_type", None),
+                        self.args.guided_decoding_backend,
+                    ))
+                if reasoning_format_for_guided_decoding is not None:
+                    # SamplingParams carries a caller-provided content
+                    # constraint, but not how this model frames reasoning and
+                    # final output. Add that model-aware framing here so
+                    # xgrammar applies the guide only to Harmony's final
+                    # channel. Other formats preserve the original guide.
+                    sampling_params.guided_decoding = (
```
Comment (brnguyen2) [inline tensorrt_llm/llmapi/llm.py:1409]: This reassigns `guided_decoding` on the caller's own `SamplingParams` object (`generate_async` doesn't copy it). After the first request, the user's `GuidedDecodingParams(json=...)` is silently replaced by an xgrammar structural-tag guide; if they reuse the same `SamplingParams` against another `LLM` (e.g. an llguidance-backed one), the structural tag — unsupported there — is what gets submitted, and their original schema is unrecoverable. The existing mutations in this method (end_id, logits processors) are additive; swapping the guide type is observably different. Consider applying the adaptation to a copy, or at least deferring the rewrite to the point where the request params are marshalled rather than the shared object.

File: tensorrt_llm/serve/openai_protocol.py [hunk 7]
```diff
@@ -422,64 +423,8 @@ def _response_format_to_guided_decoding_params(
...
-                    "end": parser.reasoning_end,
-                },
-                content,
-            ],
-        }
-
-    stag_format = ResponseFormat(type="structural_tag", format=stag_format)
-    return GuidedDecodingParams(structural_tag=stag_format.model_dump_json(
-        by_alias=True, exclude_none=True))
+    return adapt_guided_decoding_params_for_reasoning_parser(
```
Comment (brnguyen2) [inline tensorrt_llm/serve/openai_protocol.py:426]: Minor behavior change from the refactor: the old code round-tripped the built structural tag through `ResponseFormat.format`, which is typed `xgrammar.structural_tag.Format`, so the tag (including the embedded user schema/regex content) was pydantic-validated at request time and a bad one returned a 4xx. `adapt_guided_decoding_params_for_reasoning_parser` now `json.dumps` a raw dict, so malformed structural tags are only caught later in the executor. Probably acceptable since the dict layout is constant and covered by the new unit tests, but if request-time validation was intentional, consider validating the dict against `xgrammar.structural_tag.Format` inside the shared helper.

File: tests/unittest/evaluate/test_generation_content.py (hunk 5 above)
Comment (dongfengy) [reply tests/unittest/evaluate/test_generation_content.py:20]: Fixed.

**Comment by dongfengy** [comment]
> The fix looks right and the shared adapter is a real improvement over the duplicated serving logic (it also fixes the latent unbound-`content` NameError in the old `_response_format_to_guided_decoding_params` when no guide field matched). Two things should land with this PR so CI actually guards the fix:
> 
> 1. **`waives.txt` still skips the regression test.** `accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_eagle3_guided_decoding_4gpus[one_model]` is SKIPped at tests/integration/test_lists/waives.txt:59 (under https://nvbugs/5596343). Since this PR is the fix for that failure (split into 6284101) and the description says the exact 4-GPU regression passed on head, remove the waiver here — otherwise the fixed test stays disabled and nothing in CI covers the bug.
> 2. **The new unit-test file never runs in CI** — see the inline comment on `tests/unittest/evaluate/test_generation_content.py`.
> 
> One scope question worth a sentence in the description or a tracked follow-up: the raw-path adaptation is deliberately gated to xgrammar (`resolve_raw_guided_decoding_reasoning_parser` at [tensorrt_llm/llmapi/reasoning_parser.py:241](https://github.com/NVIDIA/TensorRT-LLM/pull/15741/files#diff-dc2a4b0af349317ceb68e4dedc5f9cf7a9f04b84676a9c4f59f05b0500677c30R241)), so GPT-OSS + llguidance via raw `LLM.generate()` still applies the guide to the whole transcript. Since structural tags are xgrammar-only that's understandable, but it leaves the original symptom reachable on the other backend — is a follow-up planned, or should llguidance + gpt_oss at least warn?

I have fixed waives.txt and added new tests to CI.

Regarding scope, yes, this fixes all instances of Harmony parser + xgrammar and has been verified with solid tests. Harmony + llguidance requires significantly more work and may even require architectural changes. I would suggest treating this as a feature request so that we don’t bloat this bug-fix PR.

**Review by ZhanruiSunCh — APPROVED** [review]
LGTM for infra part.

File: tests/unittest/llmapi/test_sampling_params.py [hunk 8]
```diff
@@ -201,6 +213,164 @@ def test_chat_top_logprobs_request_limit():
         )
 
 
+def test_harmony_guided_decoding_triggers_on_final_channel():
```
Comment (crazydemo) [inline tests/unittest/llmapi/test_sampling_params.py:216]: **[MINOR]** Added test functions lack type annotations

Per the repo coding guideline ("annotate every function"), the new test functions and parametrized tests (this one and those at lines 242, 261, 283, 297, 314, 330, 340, 363) should carry `-> None` and precise parameter annotations. Purely a style/consistency nit; does not affect behavior.

File: tests/unittest/evaluate/test_generation_content.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (crazydemo) [inline tests/unittest/evaluate/test_generation_content.py:1]: **[NIT]** Header is SPDX-only, missing full NVIDIA license block

Other test files in the repo carry the full NVIDIA copyright + Apache-2.0 header. This new file only has the two-line SPDX header. Add the standard block to match repo convention.
